config/configmapprovider/file_watch.go

```diff
@@ -0,0 +1,81 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configmapprovider
+
+import (
+	"context"
+	"errors"
+	"os"
+	"time"
+)
+
+var errNoOp = errors.New("no change")
+
+// watchFile sets up a watch on a filename to detect changes and calls onChange() with a suitable ChangeEvent.
+// The watch is cancelled when the context is done.
+func watchFile(ctx context.Context, filename string, onChange func(*ChangeEvent)) error {
+	var (
+		lastfi os.FileInfo
+	)
+
+	check := func() error {
+		currfi, err := os.Stat(filename)
+		modified := false
+		if err != nil {
+			if os.IsNotExist(err) && lastfi != nil {
+				return errNoOp
+			}
+			return err
```

Comment:
This appears to trigger reloading. Why do we reload if we can't stat the file?

Reply:
fixed this one. thanks for catching this. as a result, i changed the flow up slightly to ensure that if someone writes a faulty config, we preserve the last sane state as long as we dont shut down the process. Such a bug would bring down the collector across the entire kube cluster if there was a faulty config map update.